modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/infrastructure/snapshot/DataApplicationSnapshotAcceptanceManager.scala

```diff
@@ -150,42 +151,69 @@ object DataApplicationSnapshotAcceptanceManager {
         )
 
         dataState = DataState(lastOnChainState, lastCalculatedState)
+        initialResult = (
+          dataState,
+          List.empty[Signed[FeeTransaction]],
+          List.empty[Signed[DataApplicationBlock]],
+          List.empty[(Signed[DataApplicationBlock], DataBlockNotAccepted)]
+        )
 
-        (validatedUpdates, validatedBlocks, notAccepted) <- OptionT.liftF {
+        processingResult <- OptionT.liftF {
           NonEmptyList
             .fromList(dataBlocks.distinctBy(_.value.roundId))
-            .map { uniqueBlocks =>
-              uniqueBlocks.toList.traverse { dataBlock =>
+            .map(_.toList)
+            .getOrElse(Nil)
+            .foldLeftM(initialResult) {
+              case ((currentState, accFeeTransactions, accAcceptedBlocks, accNotAcceptedBlocks), dataBlock) =>
                 val dataTransactions = dataBlock.value.dataTransactions
 
                 val dataTransactionsValidations =
                   dataTransactions.traverse(validateDataTransactionsL0(_, service, balances, currentOrdinal, dataState)).map(_.reduce)
 
-                dataTransactionsValidations.flatTap { validated =>
-                  logger.warn(s"Data application is invalid, errors: ${validated.toString}").whenA(validated.isInvalid)
-                }
-                  .map(x => if (x.isInvalid) Some((dataBlock, DataBlockNotAccepted(x.toString))) else None)
-                  .handleErrorWith(err =>
-                    logger
-                      .error(err)("Unhandled exception during validating data application, assumed as invalid")
-                      .as(Some((dataBlock, DataBlockNotAccepted(err.getMessage))))
-                  )
-              }.map(_.flatten).map { invalidDataBlocks =>
-                if (invalidDataBlocks.isEmpty) {
-                  val updates = uniqueBlocks.flatMap(_.value.dataTransactions)
-                  (updates.toList, uniqueBlocks.toList, List.empty)
-                } else {
-                  (List.empty, List.empty, invalidDataBlocks)
+                dataTransactionsValidations.flatTap { validation =>
+                  if (validation.isValid)
+                    logger.debug(s"Validating block with roundId=${dataBlock.value.roundId}")
+                  else
+                    logger.debug(s"Block ${dataBlock.value.roundId} is invalid: ${validation.fold(_.toList.mkString(", "), _ => "")}")
+                }.flatMap {
+                  case Valid(_) =>
+                    val dataTransactionsAsList = dataTransactions.toList
+                    val dataUpdates = getDataUpdates(dataTransactionsAsList)
+                    val feeTransactions = getFeeTransactions(dataTransactionsAsList)
+
+                    logger.debug(s"Block ${dataBlock.value.roundId} is valid") >>
+                      service.combine(currentState, dataUpdates).map { newState =>
```

Comment:
For anyone else reviewing, this is and

`.foldLeftM(initialResult) {`

are the most important changes. We discussed this on slack already, but it wasn't properly updating the state